src/Interpreters/InterpreterParallelWithQuery.cpp

```diff
@@ -0,0 +1,180 @@
+#include <Interpreters/InterpreterParallelWithQuery.h>
+
+#include <Core/Settings.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterFactory.h>
+#include <Interpreters/executeQuery.h>
+#include <Parsers/queryToString.h>
+#include <Processors/Executors/CompletedPipelineExecutor.h>
+
+
+namespace CurrentMetrics
+{
+    extern const Metric ParallelWithQueryThreads;
+    extern const Metric ParallelWithQueryActiveThreads;
+    extern const Metric ParallelWithQueryScheduledThreads;
+}
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_QUERY;
+}
+
+namespace Setting
+{
+    extern const SettingsUInt64 parallel_with_query_max_threads;
+}
+
+
+InterpreterParallelWithQuery::InterpreterParallelWithQuery(const ASTPtr & query_, ContextPtr context_)
+    : WithContext(context_), query(query_), log(getLogger("ParallelWithQuery"))
+{
+}
+
+BlockIO InterpreterParallelWithQuery::execute()
+{
+    ASTParallelWithQuery & parallel_with_query = query->as<ASTParallelWithQuery &>();
+
+    const auto & settings = getContext()->getSettingsRef();
+    size_t max_threads = settings[Setting::parallel_with_query_max_threads];
+
+    ThreadPoolCallbackRunnerUnsafe<void> schedule;
+    std::unique_ptr<ThreadPool> thread_pool;
+
+    if (max_threads > 1)
+    {
+        thread_pool = std::make_unique<ThreadPool>(
+            CurrentMetrics::ParallelWithQueryThreads,
+            CurrentMetrics::ParallelWithQueryActiveThreads,
+            CurrentMetrics::ParallelWithQueryScheduledThreads,
+            max_threads);
+        schedule = threadPoolCallbackRunnerUnsafe<void>(*thread_pool, "ParallelWithQry");
+    }
```

Comment:
Maybe it would be better to build 1 big pipeline using https://github.com/ClickHouse/ClickHouse/blob/fe35e6d3def4c73755e7bd20b65ebe2a2e0a52a4/src/QueryPipeline/QueryPipeline.h#L133, which already run with max_threads setting? Because right not this can spawn a lot of threads.

In general, it seems to be accurate to allow any queries if the completed pipelines are built for them. For example, it will be possible to run many SELECT INTO OUTFILE.

What do you think?

Reply:
I implemented using 1 big pipeline.

Also I changed the code to allow any queries if the completed pipelines are built for them.

However it doesn't work for `SELECT INTO OUTFILE` because the server executes `SELECT INTO OUTFILE` just as a regular `SELECT`, i.e. with generating an output, and then the client writes that output into a file.